File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter13/SilentPiano/SilentPiano/Key.cpp
#include "pch.h"
#include "Key.h"

using namespace SilentPiano;

using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Data;
using namespace Windows::UI::Xaml::Documents;
using namespace Windows::UI::Xaml::Input;
using namespace Windows::UI::Xaml::Interop;
using namespace Windows::UI::Xaml::Media;

DependencyProperty^ Key::isPressedProperty = 
    DependencyProperty::Register("IsPressed", 
                                 TypeName(bool::typeid), 
                                 TypeName(Key::typeid),
                                 ref new PropertyMetadata(false, 
                                 ref new PropertyChangedCallback(OnIsPressedChanged)));

Key::Key()
{
    DefaultStyleKey = "SilentPiano.Key";
    pointerList = ref new Vector<unsigned int>();
}

DependencyProperty^ Key::IsPressedProperty::get()
{
    return isPressedProperty;
}

void Key::IsPressed::set(bool angle)
{
    SetValue(IsPressedProperty, angle);
}

bool Key::IsPressed::get()
{
    return (bool)GetValue(IsPressedProperty);
}

void Key::OnPointerEntered(PointerRoutedEventArgs^ args)
{
    if (args->Pointer->IsInContact)
        AddToList(args->Pointer->PointerId);
}

void Key::OnPointerPressed(PointerRoutedEventArgs^ args)
{
    AddToList(args->Pointer->PointerId);
}

void Key::OnPointerReleased(PointerRoutedEventArgs^ args)
{
    RemoveFromList(args->Pointer->PointerId);
}

void Key::OnPointerExited(PointerRoutedEventArgs^ args)
{
    RemoveFromList(args->Pointer->PointerId);
}

void Key::AddToList(unsigned int id)
{
    unsigned int index;
    if (!pointerList->IndexOf(id, &index))
        pointerList->Append(id);

    CheckList();
}

void Key::RemoveFromList(unsigned int id)
{
    unsigned int index;
    if (pointerList->IndexOf(id, &index))
        pointerList->RemoveAt(index);

    CheckList();
}

void Key::CheckList()
{
    this->IsPressed = pointerList->Size > 0;
}

void Key::OnIsPressedChanged(DependencyObject^ obj,
                             DependencyPropertyChangedEventArgs^ args)
{
    VisualStateManager::GoToState(dynamic_cast<Key^>(obj), 
                    (bool)args->NewValue ? "Pressed" : "Normal", false);
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter13/SilentPiano/SilentPiano/Key.cs
using System.Collections.Generic;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Input;

namespace SilentPiano
{
    public class Key : Control
    {
        static readonly DependencyProperty isPressedProperty =
                DependencyProperty.Register("IsPressed",
                        typeof(bool), typeof(Key), 
                        new PropertyMetadata(false, OnIsPressedChanged));

        List<uint> pointerList = new List<uint>();

        public static DependencyProperty IsPressedProperty 
        {
            get { return isPressedProperty; }
        }

        public bool IsPressed
        {
            set { SetValue(IsPressedProperty, value); }
            get { return (bool)GetValue(IsPressedProperty); }
        }

        protected override void OnPointerEntered(PointerRoutedEventArgs args)
        {
            if (args.Pointer.IsInContact)
                AddToList(args.Pointer.PointerId);
            base.OnPointerEntered(args);
        }

        protected override void OnPointerPressed(PointerRoutedEventArgs args)
        {
            AddToList(args.Pointer.PointerId);
            base.OnPointerPressed(args);
        }

        protected override void OnPointerReleased(PointerRoutedEventArgs args)
        {
            RemoveFromList(args.Pointer.PointerId);
            base.OnPointerReleased(args);
        }

        protected override void OnPointerExited(PointerRoutedEventArgs args)
        {
            RemoveFromList(args.Pointer.PointerId);
            base.OnPointerExited(args);
        }

        void AddToList(uint id)
        {
            if (!pointerList.Contains(id))
                pointerList.Add(id);

            CheckList();
        }

        void RemoveFromList(uint id)
        {
            if (pointerList.Contains(id))
                pointerList.Remove(id);

            CheckList();
        }

        void CheckList()
        {
            this.IsPressed = pointerList.Count > 0;
        }

        static void OnIsPressedChanged(DependencyObject obj, 
                                       DependencyPropertyChangedEventArgs args)
        {
            VisualStateManager.GoToState(obj as Key, 
                            (bool)args.NewValue ? "Pressed" : "Normal", false);
        }
    }
}
